### ISSUE TITLE
Make CLIEngine and supporting functions async via Promises

### ISSUE THREAD

**Issue by Aghassi** [eslint/eslint#10606]
**The version of ESLint you are using.**
5.1.0

**The problem you want to solve.**
CLIEngine is synchronous in nature, and I would like it to be asynchronous/promisfied

**Your take on the correct solution to problem.**
Instead of using `fs` calls that are `synchronous`, use the async ones, as well as allowing long running calls like `executeOnFiles` to return a `Promise` instead of an object. Upon resolution of the promise, pass the object via a `resolve`. This allows code using `CLIEngine` to not have to block. It also allows for things like Observables since we can use the `from` function to wrap the `Promise` and monitor output.

I also just noticed this, which seems to be in line with this request: https://www.bountysource.com/issues/26284182-lint-multiple-files-in-parallel

**Comment by not-an-aardvark** [comment]
Hi, thanks for creating an issue. As you noted, this is similar to https://github.com/eslint/eslint/issues/3565. I think it would be a good idea to have asynchronous versions of the `CLIEngine` functions, but as discussed in that issue there are a few constraints:

* The existing synchronous versions still need to be supported for compatibility.
* Linting is mostly CPU-bound rather than IO-bound, so even if it returns a `Promise` it's still going to block the event loop for awhile unless we parallelize it.

Some more discussion can be found at https://github.com/eslint/eslint/issues/3565#issuecomment-342719806.

**Comment by not-an-aardvark** [comment]
I'm not sure it's worth exposing an asynchronous API that just spawns new processes/threads if it's not using the threads to parallelize and run more quickly. An API consumer could presumably just spawn a new thread and invoke ESLint on it themselves for an equivalent benefit.

That said, I think this could be a good approach to use internally for parallelization. For example, one starting implementation of parallel `executeOnFiles` could be to create another method returning a Promise (e.g. `executeOnFilesInParallel`) that just spawns threads that each call the synchronous functions for executing on a single file. This is roughly what [`esprint`](https://github.com/pinterest/esprint) does.

(It's slightly more complicated than this because `executeOnFiles` currently doesn't just call `executeOnText` multiple times -- it also needs to handle some state that's shared among all files, e.g. for handling the set of loaded plugins and for writing cache files. It might be feasible to avoid the problem by just loading all plugins separately on each thread -- I'm not sure how much this would affect performance.)

As a minor note, creating threads would be preferable to `fork`, particularly if we want to run many of them at a time for parallelization. There are several libraries, such as [`worker-farm`](https://github.com/rvagg/node-worker-farm), to make the process easier. Node apparently supports workers natively now too, but we can't use them for awhile since we still support old versions.

**Comment by Aghassi** [comment]
@not-an-aardvark I took a brief stab at making some functions that have "async" versions. I haven't tested them (yet), but wanted to get your opinion on the direction. Would you mind having a look?
https://github.com/eslint/eslint/compare/master...Aghassi:async-execution

I didn't want to make a PR if it was throw away code.

**Comment by not-an-aardvark** [comment]
I think that looks like a good start. I notice there's a lot of code duplication from the rest of `CLIEngine`, so it would be good to eventually get rid of that where possible. For parallelization, I think we would eventually want to create a new thread for each call to `processText`.

**Comment by Aghassi** [comment]
Agreed. I think for now while the new functions are added, it is ok to have that duplicate code. Once released, create a transition state, but note up front that the synchronous versions will be phased out etc. I'll look into the thread creation for process text. I will open a PR this week with what I have so it can start getting reviewed etc. Would be nice to get the ball rolling on this 😄 . Can't wait to have faster linting!

**Comment by ilyavolodin** [comment]
> For parallelization, I think we would eventually want to create a new thread for each call to processText.

I just want to clarify here. We actually really don't want to create a new thread for each call to `processText` at all. On Windows, it's very expansive to create new threads in Node, since the basically spawns a whole new Node process. What we want to do, is create a bunch of threads first, and keep reusing them. So, something like ThreadPool.

**Comment by Aghassi** [comment]
@ilyavolodin Good to know. Do you have any examples of good implentations of that sort of thing to reference? Essentially we want a threadpool with a queue?

**Comment by Aghassi** [comment]
@not-an-aardvark @ilyavolodin Here's a question that may get this a slight speed boost without dealing with threading immediatly. What if we did diff's after first install? So, for example, let's say you do a lint on a fresh pull. That will take O(n) time because you have to do every file. However, we could use something like https://github.com/yeoman/configstore and keep a running list of all files we lint, as well as the last modified date on disk. Then, on second run, we can check if the `lastModified` field of the file has changed. If it has, we lint, otherwise, we skip. We would still take O(n) time in the worst case, but in the best case we end up with O(1) assuming only one file has changed. That would solve for general use need. Beyond that, we can also look to make the rest of process asynchronous.

**Comment by platinumazure** [comment]
Hi @Aghassi, have you checked out the `--cache` CLI option and/or the `cache: true` CLIEngine constructor option? This does most of what you are suggesting :smile:

**Comment by not-an-aardvark** [comment]
That already exists -- see the `--cache` flag.

**Comment by Aghassi** [comment]
@platinumazure Was not aware! I will check that out. Ok so regardless, we still have the threading/async issue 😅

**Comment by FreezePhoenix** [comment]
So this is still an ongoing project. I have just one thing to add: This should have been part of it before eslint got too large to manage.

**Comment by not-an-aardvark** [comment]
@FreezePhoenix Yes, parallel linting is still not implemented yet. I think the question of whether ESLint should have implemented a different API earlier is moot at this point.

**Comment by FreezePhoenix** [comment]
@not-an-aardvark  I'm not saying they did something wrong - I'm saying that would have been a better idea.

**Comment by Aghassi** [comment]
@not-an-aardvark Node standard API has sync and a sync API. So does Babel. I think it is a common practice and good to offer both.

**Comment by remcohaszing** [comment]
Does this mean rules can be async too? E.g.

```jsx
module.exports = {
  meta: {
    // ...
  },

  create(context) {
    return {
      async Program(node) {
        await thirdPartyLibrary(node);
      }
    }
  }
}
```

I’m trying to build a plugin that heavily relies on a third party library, but I’m stuck because they only support async calls, whereas this is not supported by ESLint.

**Comment by not-an-aardvark** [comment]
@remcohaszing I don't think async rules are currently planned, but if we do eventually support async rules, then making `CLIEngine` async would be a prerequisite.

**Comment by FreezePhoenix** [comment]
Or... we can just rewrite the entire thing.
